File: summary/ts/18_half_enumeration.ts
function halfEnumeration(arr: number[], targetSum: number): number {
  const n = arr.length;
  const half = Math.floor(n / 2);

  const firstHalfSubsets: { [key: number]: number } = {};
  for (let i = 0; i < (1 << half); i++) {
    let sum = 0;
    for (let j = 0; j < half; j++) {
      if (i & (1 << j)) {
        sum += arr[j];
      }
    }
    firstHalfSubsets[sum] = (firstHalfSubsets[sum] || 0) + 1;
  }

  const secondHalfSubsets: { [key: number]: number } = {};
  for (let i = 0; i < (1 << (n - half)); i++) {
    let sum = 0;
    for (let j = 0; j < (n - half); j++) {
      if (i & (1 << j)) {
        sum += arr[half + j];
      }
    }
    secondHalfSubsets[sum] = (secondHalfSubsets[sum] || 0) + 1;
  }

  let count = 0;
  for (const [sum1, count1] of Object.entries(firstHalfSubsets)) {
    const needed = targetSum - parseInt(sum1);
    count += count1 * (secondHalfSubsets[needed] || 0);
  }

  return count;
}

// Example usage:
const inputArray = [1, 2, 3, 4, 5, 6, 7, 8];
const targetSum = 10;
const result = halfEnumeration(inputArray, targetSum);
console.log(`Number of ways to achieve sum ${targetSum}: ${result}`);

File: summary/dart/18_half_enumeration.dart
import 'dart:io';

int halfEnumeration(List<int> arr, int targetSum) {
  int n = arr.length;
  int half = n ~/ 2;

  Map<int, int> firstHalfSubsets = {};
  for (int i = 0; i < (1 << half); i++) {
    int sum = 0;
    for (int j = 0; j < half; j++) {
      if ((i & (1 << j)) != 0) {
        sum += arr[j];
      }
    }
    firstHalfSubsets[sum] = (firstHalfSubsets[sum] ?? 0) + 1;
  }

  Map<int, int> secondHalfSubsets = {};
  for (int i = 0; i < (1 << (n - half)); i++) {
    int sum = 0;
    for (int j = 0; j < (n - half); j++) {
      if ((i & (1 << j)) != 0) {
        sum += arr[half + j];
      }
    }
    secondHalfSubsets[sum] = (secondHalfSubsets[sum] ?? 0) + 1;
  }

  int count = 0;
  firstHalfSubsets.forEach((sum1, count1) {
    int needed = targetSum - sum1;
    count += count1 * (secondHalfSubsets[needed] ?? 0);
  });

  return count;
}

void main() {
  print("Enter the array elements separated by space:");
  String? inputArray = stdin.readLineSync();
  if (inputArray != null) {
    List<int> arr = inputArray.split(' ').map(int.parse).toList();
    print("Enter the target sum:");
    String? inputTarget = stdin.readLineSync();
    if (inputTarget != null) {
      int targetSum = int.parse(inputTarget);
      int result = halfEnumeration(arr, targetSum);
      print("Number of ways to achieve sum $targetSum: $result");
    }
  }
}
